packages/block-library/src/navigation/index.php

```diff
@@ -67,6 +67,521 @@ function block_core_navigation_sort_menu_items_by_parent_id( $menu_items ) {
+		 */
+
+		$has_old_responsive_attribute = ! empty( $attributes['isResponsive'] ) && $attributes['isResponsive'];
+		return isset( $attributes['overlayMenu'] ) && 'never' !== $attributes['overlayMenu'] || $has_old_responsive_attribute;
+	}
+
+	/**
+	 * Returns whether or not a navigation has a submenu.
+	 */
+	private static function does_navigation_have_submenus( $inner_blocks ) {
+		$has_submenus = false;
+		foreach ( $inner_blocks as $inner_block ) {
+			$inner_block_content = $inner_block->render();
+			$p                   = new WP_HTML_Tag_Processor( $inner_block_content );
+			if ( $p->next_tag(
+				array(
+					'name'       => 'LI',
+					'class_name' => 'has-child',
+				)
+			) ) {
+				$has_submenus = true;
+			}
+		}
+		return $has_submenus;
+	}
+
+	private static function should_load_view_script( $attributes, $inner_blocks ) {
+		$has_submenus = WP_Navigation_Block::does_navigation_have_submenus( $inner_blocks );
+		$is_responsive_menu = WP_Navigation_Block::is_responsive_navigation( $attributes );
+		return ( $has_submenus && ( $attributes['openSubmenusOnClick'] || $attributes['showSubmenuIcon'] ) ) || $is_responsive_menu;
+	}
+
+	/**
+	 * Returns the html for the inner blocks of the navigation block.
+	 */
+	private static function get_inner_blocks_html( $inner_blocks, $attributes ) {
+		$has_submenus            = WP_Navigation_Block::does_navigation_have_submenus( $inner_blocks );
+		$should_load_view_script = WP_Navigation_Block::should_load_view_script( $attributes, $inner_blocks );
+
+		$list_item_nav_blocks = array(
+			'core/navigation-link',
+			'core/home-link',
+			'core/site-title',
+			'core/site-logo',
+			'core/navigation-submenu',
+		);
+
+		$needs_list_item_wrapper = array(
+			'core/site-title',
+			'core/site-logo',
+		);
+
+		$style = WP_Navigation_Block::get_styles( $attributes );
+		$class = WP_Navigation_Block::get_classes( $attributes );
+		$container_attributes = get_block_wrapper_attributes(
+			array(
+				'class' => 'wp-block-navigation__container ' . $class,
+				'style' => $style,
+			)
+		);
+
+		$inner_blocks_html = '';
+		$is_list_open      = false;
+
+		foreach ( $inner_blocks as $inner_block ) {
+			$is_list_item = in_array( $inner_block->name, $list_item_nav_blocks, true );
+
+			if ( $is_list_item && ! $is_list_open ) {
+				$is_list_open       = true;
+				$inner_blocks_html .= sprintf(
+					'<ul %1$s>',
+					$container_attributes
+				);
+			}
+
+			if ( ! $is_list_item && $is_list_open ) {
+				$is_list_open       = false;
+				$inner_blocks_html .= '</ul>';
+			}
+
+			$inner_block_content = $inner_block->render();
+			if ( ! empty( $inner_block_content ) ) {
+				if ( in_array( $inner_block->name, $needs_list_item_wrapper, true ) ) {
+					$inner_blocks_html .= '<li class="wp-block-navigation-item">' . $inner_block_content . '</li>';
+				} else {
+					$inner_blocks_html .= $inner_block_content;
+				}
```

Comment:
This part of the method almost feels that it would be better as a separate method. Why? Because then you could unit test this code around wrapping and accessibility.

I really think we need tests on this part.

Reply:
I'm not sure we can separate this part without passing a lot of parameters to it. The logic requires the use of the following variables which are created earlier in the method:

`$container_attributes`, `$class`, `$style`, `$needs_list_item_wrapper`, `$list_item_nav_blocks`.

We could remove the end part which adds the directives if you felt that should be separate.

Reply:
The bit that needs test coverage is the code which determines whether to wrap with lists or not. It's prone to error.

At least two of the parameters you mention could be extracted to static members on the class rather than being local to the method. That leaves 3 params which doesn't feel excessive.

I think it would be worth it to slim down the parent method which is currently doing too much and also to provide more confidence about this particularly unusual piece of code.

Reply:
I added a new function that just handles this logic in 103c27ab2d